app/src/main/java/org/mozilla/fenix/home/PocketUpdatesMiddleware.kt

```diff
@@ -4,25 +4,63 @@
 
 package org.mozilla.fenix.home
 
+import androidx.datastore.core.DataStore
 import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.collect
 import kotlinx.coroutines.launch
 import mozilla.components.lib.state.Action
 import mozilla.components.lib.state.Middleware
 import mozilla.components.lib.state.MiddlewareContext
 import mozilla.components.service.pocket.PocketStoriesService
+import org.mozilla.fenix.datastore.SelectedPocketStoriesCategories
+import org.mozilla.fenix.datastore.SelectedPocketStoriesCategories.SelectedPocketStoriesCategory
+import org.mozilla.fenix.home.sessioncontrol.viewholders.pocket.PocketRecommendedStoriesSelectedCategory
 
 /**
  * [HomeFragmentStore] middleware reacting in response to Pocket related [Action]s.
+ *
+ * @param coroutineScope [CoroutineScope] used for long running operations like disk IO.
+ * @param pocketStoriesService [PocketStoriesService] used for updating details about the Pocket recommended stories.
+ * @param selectedPocketCategoriesStore [DataStore] used for reading or persisting details about the
+ * currently selected Pocket recommended stories categories.
  */
 class PocketUpdatesMiddleware(
     private val coroutineScope: CoroutineScope,
-    private val pocketStoriesService: PocketStoriesService
+    private val pocketStoriesService: PocketStoriesService,
+    private val selectedPocketCategoriesStore: DataStore<SelectedPocketStoriesCategories>
 ) : Middleware<HomeFragmentState, HomeFragmentAction> {
     override fun invoke(
         context: MiddlewareContext<HomeFragmentState, HomeFragmentAction>,
         next: (HomeFragmentAction) -> Unit,
         action: HomeFragmentAction
     ) {
+        // Pre process actions
+        when (action) {
+            is HomeFragmentAction.PocketStoriesCategoriesChange -> {
+                // Whenever the stories categories are updated we need to restore their "selected" status.
+
+                coroutineScope.launch {
+                    selectedPocketCategoriesStore.data.collect { persistedSelectedCategories ->
+                        // Replace the original action with one that also
+                        context.store.dispatch(
+                            HomeFragmentAction.PocketStoriesCategoriesSelectionsChange(
+                                action.storiesCategories,
+                                persistedSelectedCategories.valuesList.map {
+                                    PocketRecommendedStoriesSelectedCategory(
+                                        name = it.name,
+                                        selectionTimestamp = it.selectionTimestamp
+                                    )
+                                }
+                            )
+                        )
+                    }
```

Comment:
nit: I tend to move this code that launches coroutines to private functions in the same file and then pass only the data/references they need to them. The advantage is that this code will no longer have access to the `MiddlewareContext` which only makes sense in the thread of the Middleware itself. For example see `EngineDelegateMiddleware`.